source/opt/instruction.h

```diff
@@ -506,28 +512,27 @@ inline void Instruction::SetInOperands(OperandList&& new_operands) {
 }
 
 inline void Instruction::SetResultId(uint32_t res_id) {
-  result_id_ = res_id;
-  auto ridx = (type_id_ != 0) ? 1 : 0;
+  auto ridx = has_type_id_ ? 1 : 0;
   assert(operands_[ridx].type == SPV_OPERAND_TYPE_RESULT_ID);
   operands_[ridx].words = {res_id};
 }
 
 inline void Instruction::SetResultType(uint32_t ty_id) {
-  if (type_id_ != 0) {
-    type_id_ = ty_id;
-    assert(operands_.front().type == SPV_OPERAND_TYPE_TYPE_ID);
-    operands_.front().words = {ty_id};
-  }
+  if (!has_type_id_) return;
```

Comment:
This preserves behaviour:  an instruction has a type id if and only if it always had a type. 
It's a bit weird because you can't take a Nop instruction and turn it into something like OpConstant.  You have to construct an instruction with a type if it ever needs a type.
But that's all a pre-existing problem.

Reply:
I think the issue is adding and removing things from the Operand list if you add/remove type and result ids.

Reply:
Yes there is a problem that the existence of a type and result id should be determined by the opcode, but we get problems with consistency, but for this PR just retain the same behaviour.

With that in mind, if we set the result type to "0" in the old implementation, then later checks would think there was no type id.  Then the operands would not be lined up correctly.  This was incorrect.  

I think we should document and assert that you cannot change the result type to 0.

Reply:
Done.